### PR TITLE
MAKE-1240: Implement the HistoricalTestData model

### PR DESCRIPTION
**Pull request by julianedwards**

### CODE REVIEW

File: model/historical_test_data.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	LastUpdate      time.Time `bson:"last_update"`
+
+	env       cedar.Environment
+	populated bool
+}
+
+// CreateHistoricalTestData is an entry point for creating a new
+// HistoricalTestData.
+func CreateHistoricalTestData(info HistoricalTestDataInfo) (*HistoricalTestData, error) {
+	if err := info.validate(); err != nil {
```
Comment (tychoish) [inline model/historical_test_data.go:26]: why is this private?

File: model/historical_test_data.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	Project   string    `bson:"project"`
+	Variant   string    `bson:"variant"`
+	TaskName  string    `bson:"task_name"`
+	TestName  string    `bson:"test_name"`
+	Requester string    `bson:"requester"`
+	Date      time.Time `bson:"date"`
+}
+
+func (i *HistoricalTestDataInfo) validate() error {
+	var emptyField bool
```
Comment (tychoish) [inline model/historical_test_data.go:49]: I'd do a catcher with NewWhen calls and the conditionals so the error message makes it clear what's missing.

File: model/historical_test_data.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// HistoricalTestDataInfo describes information unique to a single test
+// statistics document.
+type HistoricalTestDataInfo struct {
+	Project   string    `bson:"project"`
+	Variant   string    `bson:"variant"`
+	TaskName  string    `bson:"task_name"`
+	TestName  string    `bson:"test_name"`
+	Requester string    `bson:"requester"`
```
Comment (tychoish) [inline model/historical_test_data.go:44]: requester is a weird evergreen term that maybe we want to elide/cover up?

File: model/historical_test_data.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/pkg/errors"
+)
+
+// HistoricalTestData describes aggregated test result data for a given date
+// range.
+type HistoricalTestData struct {
+	Info            HistoricalTestDataInfo
+	NumPass         int       `bson:"num_pass"`
+	NumFail         int       `bson:"num_fail"`
+	AverageDuration float64   `bson:"average_duration"`
```
Comment (tychoish) [inline model/historical_test_data.go:16]: do we want to store total time in some units? also units are unclear.

File: model/historical_test_data.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	LastUpdate      time.Time `bson:"last_update"`
+
+	env       cedar.Environment
```
Comment (tychoish) [inline model/historical_test_data.go:19]: I don't think you're using this (yet? ever?)

File: model/historical_test_data.go (hunk 4 above)
Comment (julianedwards) [reply model/historical_test_data.go:16]: Would you rather store totalTime and calculate average duration on the fly using NumPass + NumFail?

File: model/historical_test_data.go (hunk 5 above)
Comment (julianedwards) [reply model/historical_test_data.go:19]: I will because I need to get the configuration to get the test results bucket

Comment (julianedwards) [reply model/historical_test_data.go:19]: I broke this into small tickets as a productivity hackl

File: model/historical_test_data.go (hunk 1 above)
Comment (julianedwards) [reply model/historical_test_data.go:26]: it doesn't have to be but I didn't see the need to make it public?

File: model/historical_test_data.go (hunk 3 above)
Comment (julianedwards) [reply model/historical_test_data.go:44]: its used as part of the identity... would you rather call it something else?

Comment (julianedwards) [reply model/historical_test_data.go:44]: I don't think we can get rid of it though

File: model/historical_test_data.go (hunk 4 above)
Comment (tychoish) [reply model/historical_test_data.go:16]: Just to clarify, this is going to be stored in s3? If that's true (and it probably ought to be?) then we should probably store everything (components and average?)

File: model/historical_test_data.go (hunk 5 above)
Comment (tychoish) [reply model/historical_test_data.go:19]: nocie.

File: model/historical_test_data.go (hunk 3 above)
Comment (tychoish) [reply model/historical_test_data.go:44]: We should definitely call it something else, and maybe have a maineline boolean in addition to  some better kind of name.

File: model/historical_test_data.go (hunk 4 above)
Comment (julianedwards) [reply model/historical_test_data.go:16]: Don't think there is anything else to store here? Average will be calculated as test results come in, don't see a compelling reason to store each individual test duration.

Comment (julianedwards) [reply model/historical_test_data.go:16]: Unless we are planning to do more complex rollups?

File: model/historical_test_data.go (hunk 3 above)
Comment (julianedwards) [reply model/historical_test_data.go:44]: What is a good substitute? `request_type`? Historical test statistics uses requester as part of the id and its not binary (like mainline), so we need to keep an equivalent field I think.

**Pull request merged**